### ISSUE TITLE
False positive: `unused_must_use` for boxed function

### ISSUE THREAD

**Issue by alex-ozdemir** [rust-lang/rust#106957]
### Code

```Rust
fn main() {
    let mut xs: Vec<Box<dyn Fn()>> = Vec::new();
    xs.push(Box::new(|| println!("I've been used!")));
    for x in &xs {
        x();
    }
    xs.remove(0);
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
warning: unused boxed `Fn` trait object that must be used
 --> src/main.rs:7:5
  |
7 |     xs.remove(0);
  |     ^^^^^^^^^^^^^
  |
  = note: closures are lazy and do nothing unless called
  = note: `#[warn(unused_must_use)]` on by default

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.65s
     Running `target/debug/playground`
```

### Desired output

```Shell
The function is run, so I don't think there should be a warning here.
```

### Anything else?

I wasn't sure whether this is should be a bug report or a diagnostics bug report. Sorry if I chose the wrong one!

**Comment by chenyukang** [comment]
Our current lint for `unused_must_use` is by expr, so it has a limitation that can not cpature the part before the current expr.

I think it make sense, otherwise it's need to much effort to analyze the detailed behavior of program, for example if we make the code like this:

```rust
fn main() {
    let mut xs: Vec<Box<dyn Fn()>> = Vec::new();
    xs.push(Box::new(|| println!("I've been used!")));
    xs.push(Box::new(|| println!("I've been used, two!")));
    xs[0]();
    xs.remove(1);
}
```

Now the lint is valid, since the second closure haven't been called:

```console
warning: unused boxed `Fn` trait object that must be used
  --> ./p/unused.rs:11:5
   |
11 |     xs.remove(1);
   |     ^^^^^^^^^^^^
   |
   = note: closures are lazy and do nothing unless called
   = note: `#[warn(unused_must_use)]` on by default
```

**Comment by chenyukang** [comment]
The current expr `xs.remove(0)` report the lint message, because `remove` has a return value,
We need to explicitly ignore it:

```rust
let _ = xs.remove(0);
```

**Comment by alex-ozdemir** [comment]
Thank you for looking at this!

Ah yes, I see that you're right.

I thought I might be hitting some special-case lint for closures, but now I see that this is just an instance of a `must_use` *trait*. The relevant commit is [here](https://github.com/rust-lang/rust/commit/821d50aa0cc73c1ec2fb09bab8a69e1c3ca93f23), the modern code is [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/unused.rs), and the documentation is [here](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute).

Since I imagine that this lint is really designed for cases where a closure is constructed and immediately dropped, example:
```rust
fn main() {
    || { println!("Hi"); };
    println!("{}", 0);
}
```
, we could try to be a little more restrictive. For example, in [this case](https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/unused.rs#L319), we could check whether `expr` actually contains a closure expression (i.e., and expression that is syntactically a closure).

But, that might have some unintended consequences. I'm not sure I understand all the things that we want this lint to catch.

I'll close this.
